File: taskflow-frontend/src/types/Achievement.ts
export const AchievementLevel = {
    BRONZE: 'bronze',
    SILVER: 'silver',
    GOLD: 'gold',
    DIAMOND: 'diamond'
} as const;

export type AchievementLevel = typeof AchievementLevel[keyof typeof AchievementLevel];

export const AchievementCategory = {
    PRODUCTIVITY: 'productivity',
    ORGANIZATION: 'organization',
    CONSISTENCY: 'consistency',
    EXPLORATION: 'exploration',
    MASTERY: 'mastery',
} as const;

export type AchievementCategory = typeof AchievementCategory[keyof typeof AchievementCategory];

export const AchievementType = {
    COUNTER: 'counter',
    STREAK: 'streak',
    MILESTONE: 'milestone',
    PERCENTAGE: 'percentage'
} as const;

export type AchievementType = typeof AchievementType[keyof typeof AchievementType];

export interface AchievementTier {
    level: AchievementLevel;
    target: number;
    points: number;
    unlocked: boolean;
    unlockedAt?: Date;
}

export interface Achievement {
    id: string;
    key: string;
    category: AchievementCategory;
    type: AchievementType;
    icon: string;
    color: string;
    tiers: AchievementTier[];
    isHidden: boolean;
    prerequisiteAchievements?: string[];
}

export interface AchievementProgress {
    achievementId: string;
    currentValue: number;
    lastUpdated: Date;
    streakCount?: number;
    lastStreakDate?: Date;
    unlockedTiers: AchievementLevel[];
}

export interface AchievementEvent {
    type: AchievementEventType;
    data?: any;
    timestamp: Date;
}

export const AchievementEventType = {
    TASK_CREATED: 'task_created' as const,
    TASK_COMPLETED: 'task_completed' as const,
    TASK_DELETED: 'task_deleted' as const,
    TASK_UPDATED: 'task_updated' as const,

    CATEGORY_CREATED: 'category_created' as const,
    CATEGORY_UPDATED: 'category_updated' as const,
    CATEGORY_DELETED: 'category_deleted' as const,

    APP_OPENED: 'app_opened' as const,
    CALENDAR_VIEWED: 'calendar_viewed' as const,
    DASHBOARD_VIEWED: 'dashboard_viewed' as const,

    DAILY_LOGIN: 'daily_login' as const,
    TASK_COMPLETED_ON_TIME: 'task_completed_on_time' as const,
    TASK_COMPLETED_LATE: 'task_completed_late' as const,

    ALL_TASKS_COMPLETED_TODAY: 'all_tasks_completed_today' as const,
    WEEKEND_PRODUCTIVITY: 'weekend_productivity' as const,
    EARLY_BIRD: 'early_bird' as const,
    NIGHT_OWL: 'night_owl' as const
} as const;

export type AchievementEventType = typeof AchievementEventType[keyof typeof AchievementEventType];

export interface UserAchievementStats {
    totalPoints: number;
    totalAchievements: number;
    unlockedAchievements: number;
    currentStreak: number;
    longestStreak: number;
    level: number;
    experiencePoints: number;
    nextLevelPoints: number;
}

export const USER_LEVELS = [
    { level: 1, minPoints: 0, titleKey: 'achievementsDashboard.beginner' },
    { level: 2, minPoints: 100, titleKey: 'achievementsDashboard.organized' },
    { level: 3, minPoints: 300, titleKey: 'achievementsDashboard.productive' },
    { level: 4, minPoints: 600, titleKey: 'achievementsDashboard.efficient' },
    { level: 5, minPoints: 1000, titleKey: 'achievementsDashboard.expert' },
    { level: 6, minPoints: 1500, titleKey: 'achievementsDashboard.master' },
    { level: 7, minPoints: 2200, titleKey: 'achievementsDashboard.guru' },
    { level: 8, minPoints: 3000, titleKey: 'achievementsDashboard.legend' }
];

export const calculateUserLevel = (totalPoints: number) => {
    const currentLevel = USER_LEVELS
        .slice()
        .reverse()
        .find(level => totalPoints >= level.minPoints) || USER_LEVELS[0];

    const currentLevelIndex = USER_LEVELS.findIndex(l => l.level === currentLevel.level);
    const nextLevel = USER_LEVELS[currentLevelIndex + 1];

    const experiencePoints = totalPoints - currentLevel.minPoints;
    const nextLevelPoints = nextLevel ? nextLevel.minPoints - currentLevel.minPoints : 0;

    return {
        level: currentLevel.level,
        titleKey: currentLevel.titleKey,
        experiencePoints,
        nextLevelPoints,
        progress: nextLevelPoints > 0 ? (experiencePoints / nextLevelPoints) * 100 : 100
    };
};

export interface AchievementNotification {
    achievement: Achievement;
    tier: AchievementTier;
    isNewAchievement: boolean; // true if first tier unlocked, false if tier upgrade
}

File: TaskFlow.Api/Models/Achievement.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace TaskFlow.Api.Models
{
    public class Achievement
    {
        [Key]
        public string Id { get; set; } = "";

        [Required]
        public string Key { get; set; } = "";

        [Required]
        public string Category { get; set; } = "";

        [Required]
        public string Type { get; set; } = "";

        [Required]
        public string Icon { get; set; } = "";

        public string Color { get; set; } = "";

        public bool IsHidden { get; set; } = false;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public virtual ICollection<AchievementTier> Tiers { get; set; } = new List<AchievementTier>();
        public virtual ICollection<UserAchievementProgress> UserProgress { get; set; } = new List<UserAchievementProgress>();
    }

    public class AchievementTier
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public string AchievementId { get; set; } = "";

        [Required]
        public string Level { get; set; } = ""; // bronze, silver, gold, diamond

        public int Target { get; set; }

        public int Points { get; set; }

        public int SortOrder { get; set; }

        public virtual Achievement Achievement { get; set; } = null!;
        public virtual ICollection<UserAchievementTierProgress> UserTierProgress { get; set; } = new List<UserAchievementTierProgress>();
    }

    public class UserAchievementProgress
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public int UserId { get; set; }

        [Required]
        public string AchievementId { get; set; } = "";

        public int CurrentValue { get; set; } = 0;

        public DateTime LastUpdated { get; set; } = DateTime.UtcNow;

        public DateTime? FirstUnlockedAt { get; set; }

        public virtual User User { get; set; } = null!;
        public virtual Achievement Achievement { get; set; } = null!;
        public virtual ICollection<UserAchievementTierProgress> TierProgress { get; set; } = new List<UserAchievementTierProgress>();
    }

    public class UserAchievementTierProgress
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public int UserAchievementProgressId { get; set; }

        [Required]
        public int AchievementTierId { get; set; }

        public bool IsUnlocked { get; set; } = false;

        public DateTime? UnlockedAt { get; set; }

        public virtual UserAchievementProgress UserAchievementProgress { get; set; } = null!;
        public virtual AchievementTier AchievementTier { get; set; } = null!;
    }
    public class AchievementEvent
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public int UserId { get; set; }

        [Required]
        public string EventType { get; set; } = "";

        public string? EventData { get; set; }

        public DateTime Timestamp { get; set; } = DateTime.UtcNow;

        public bool IsProcessed { get; set; } = false;

        public virtual User User { get; set; } = null!;
    }

    public class UserAchievementStats
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public int UserId { get; set; }

        public int TotalPoints { get; set; } = 0;

        public int TotalAchievements { get; set; } = 0;

        public int UnlockedAchievements { get; set; } = 0;

        public int CurrentStreak { get; set; } = 0;

        public int LongestStreak { get; set; } = 0;

        public DateTime? LastStreakDate { get; set; } = null;

        public int Level { get; set; } = 1;

        public int ExperiencePoints { get; set; } = 0;

        public int NextLevelPoints { get; set; } = 100;

        public DateTime LastUpdated { get; set; } = DateTime.UtcNow;

        public virtual User User { get; set; } = null!;
    }
}